src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/authorization/ServiceProviderAccessScopeMapper.kt

```diff
@@ -34,63 +40,81 @@ class ServiceProviderAccessScopeMapper(
     val groups = hmppsAuthService.getUserGroups(user)
       ?: throw AccessError(user, errorMessage, listOf("cannot find user in hmpps auth"))
 
-    val configErrors = mutableListOf<String>()
+    // order is important as each step can mutate WorkingScope
+    val workingScope = WorkingScope(authGroups = groups)
+
+    resolveProviders(workingScope)
+    resolveContracts(workingScope)
+    removeInaccessibleContracts(workingScope)
+
+    blockUsersWithoutProviders(workingScope)
+    blockUsersWithoutContracts(workingScope)
+
+    if (workingScope.errors.isNotEmpty()) {
+      throw AccessError(user, errorMessage, workingScope.errors)
+    }
+
+    return ServiceProviderAccessScope(
+      serviceProviders = workingScope.providers,
+      contracts = workingScope.contracts,
+    )
+  }
 
-    val serviceProviderGroups = groups
+  private fun resolveProviders(scope: WorkingScope) {
+    val serviceProviderGroups = scope.authGroups
       .filter { it.startsWith(serviceProviderGroupPrefix) }
       .map { it.removePrefix(serviceProviderGroupPrefix) }
 
-    val serviceProvider = getServiceProvider(serviceProviderGroups, configErrors)
+    val providers = getProviders(serviceProviderGroups, scope.errors)
+    scope.providers.addAll(providers)
+  }
 
-    val contractGroups = groups
+  private fun resolveContracts(scope: WorkingScope) {
+    val contractGroups = scope.authGroups
       .filter { it.startsWith(contractGroupPrefix) }
       .map { it.removePrefix(contractGroupPrefix) }
 
-    val contracts = getContracts(contractGroups, configErrors)
+    val contracts = getContracts(contractGroups, scope.errors)
+    scope.contracts.addAll(contracts)
+  }
 
-    // FIXME we also need to remove contracts which do not belong to the user's provider
+  private fun removeInaccessibleContracts(scope: WorkingScope) {
+    val orphanContracts = scope.contracts.filterNot {
+      scope.providers.contains(it.primeProvider) || it.subcontractorProviders.intersect(scope.providers).isNotEmpty()
```

Comment:
i think would be easier to read with `it` being named and without the extra `not` to grep

```
val contractsNotAssosciatedWithUserProviders = contracts.filter { contract ->
      val validProviders = setOf(contract.primeProvider) + contract.subcontractorProviders
      scope.providers.intersect(validProviders).isNotEmpty()
}
```

Reply:
done in a897e8b7f41e739e9787cb9262742d68e990ee12